### PR TITLE
Add notifications snippets in views for web based notifications

### PR DESCRIPTION
**Pull request by shreyansh23**

### CODE REVIEW

**Comment by shreyansh23** [comment]
Before testing pull https://github.com/IMGIITRoorkee/omniport-service-notifications/pull/3 to avoid `unique_together` related error.

**Comment by shreyansh23** [comment]
> Why do you require this?

@ayushjn20 I am unable to understand for which lines does think comment points.

File: utils.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import os
+import yaml
+from categories.models.category import Category
+
+
+def get_base_category():
+    """
+    Function to get the category instance with slug
```
Comment (dhruvkb) [inline utils.py]: Documentation of function should be imperative. Examples:
- Get X
- Calculate Y
- Do Z

File: utils.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    equal to the service name
+    """
+    config_path = os.path.join('services', 'helpcentre', 'config.yml')
```
Comment (dhruvkb) [inline utils.py]: Line between code and docstrings.

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    with open(config_path, 'r') as f:
```
Comment (dhruvkb) [inline utils.py]: I'm sure there are most descriptive variables names than `f`.

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        app_configuration = yaml.safe_load(f)
+    service_name = app_configuration['nomenclature']['name']
+    service_verbose = app_configuration['nomenclature']['verboseName']
+    cat_obj, created = Category.objects.get_or_create(
+        slug=service_name, name=service_verbose)
```
Comment (dhruvkb) [inline utils.py]: Line break between arguments, closing parenthesis must be on its own line.

File: views/views.py [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import swapper
 from rest_framework import status
```
Comment (dhruvkb) [inline views/views.py]: Blank line between imports of different packages.

File: views/views.py [hunk 6]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
                     },
                     status=status.HTTP_403_FORBIDDEN,
                 )
+
+        """ Notifications start """
```
Comment (dhruvkb) [inline views/views.py]: Use `#` for comments that are not docstrings.

File: views/views.py [hunk 7]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
...
+                                  person=person_id, is_personalised=True)
+
+        # For assignee related notification
+        if 'assignees' in request.data.keys():
+            new_assignees = request.data['assignees']
+            old_assignees = list(query.assignees.all().values_list(
+                'id', flat=True))
+            is_added = None
+            assignee_id = None
+            for i in new_assignees:
```
Comment (dhruvkb) [inline views/views.py]: `i`?

File: views/views.py [hunk 8]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
...
+                if i not in new_assignees:
+                    is_added = False
+                    assignee_id = i
+                    break
+            status = "assigned to" if is_added is True else "unassigned from"
+            common_assignees = [i for i in new_assignees if i in old_assignees]
+
+            # For the fellow assignees
+            person = Maintainer.objects.get(id=assignee_id).person
+            person_ids = [Maintainer.objects.get(
```
Comment (dhruvkb) [inline views/views.py]: Indentation style of Omniport is as follows.
```python
x = [
    ...
    ...
]
```

File: views/views.py [hunk 9]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
...
+                id=i).person_id for i in common_assignees]
```
Comment (dhruvkb) [inline views/views.py]: Why did you break line here, `Maintainer.objects.get(id=i).person_id` would be fine on one line.

File: views/views.py [hunk 10]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
...
+            person_ids = [i for i in person_ids if i != request_person.id]
+            template = f"{person.full_name} has been {status} query "\
```
Comment (dhruvkb) [inline views/views.py]: Use `' ... '` in Omniport.

Comment (dhruvkb) [inline views/views.py]: If strings are long use brackets instead of `\`.
```python
x = (
    f'...'
    f'...'
)
```

File: views/views.py [hunk 11]
```diff
@@ -88,6 +96,85 @@ def partial_update(self, request, *args, **kwargs):
...
+                f"{query.title} opened by {query.uploader.full_name}"\
+                f" for app {query.app_name}"
+            if person_ids:
+                push_notification(template=template, category=category_obj,
```
Comment (dhruvkb) [inline views/views.py]:
```suggestion
                push_notification(
                    template=template,
                    category=category_obj,
                    ...
```

**Comment by dhruvkb** [comment]
@shreyansh23 can you mark the changes that you made as resolved? I'm from a mobile device so that would make reviewing much easier.

**Comment by shreyansh23** [comment]
>  a mobile device so that would make reviewing much easier.
@dhruvkb Yes, I will do.

File: views/views.py (hunk 8 above)
Comment (shreyansh23) [reply views/views.py]: Yes, followed this in the updated code.

File: views/views.py (hunk 10 above)
Comment (shreyansh23) [reply views/views.py]: Used single quotes for strings in updated code.

File: views/views.py (hunk 11 above)
Comment (shreyansh23) [reply views/views.py]: Followed this now. :+1:

File: utils.py (hunk 4 above)
Comment (shreyansh23) [reply utils.py]: Added :+1:

File: utils.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import os
+import yaml
+
```
Comment (dhruvkb) [inline utils.py]: Two lines after import.

File: utils.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+from categories.models.category import Category
+
+
+def get_base_category():
+    """
```
Comment (dhruvkb) [inline utils.py]: Documentation is missing.

File: views/views.py [hunk 14]
```diff
@@ -7,6 +9,13 @@
 from helpcentre.serializers.serializers import *
 from kernel.utils.rights import has_helpcentre_rights
 from kernel.managers.get_role import get_role
+from notifications.actions import push_notification
```
Comment (dhruvkb) [inline views/views.py]: Alphabetical order of imports.

File: views/views.py [hunk 15]
```diff
@@ -88,6 +97,112 @@ def partial_update(self, request, *args, **kwargs):
...
+                )
+                push_notification(
+                    template=template,
+                    category=category_obj,
+                    web_onclick_url=url,
+                    person=person_id,
+                    is_personalised=True
+                )
+
+        # Notifications end
```
Comment (dhruvkb) [inline views/views.py]: Never do this.

File: views/views.py [hunk 16]
```diff
@@ -147,6 +262,50 @@ def create(self, request, *args, **kwargs):
...
+                )
+                push_notification(
+                    template=template,
+                    category=category_obj,
+                    web_onclick_url=url,
+                    person=uploader_id,
+                    is_personalised=True
+                )
+
+            # Notifications end
```
Comment (dhruvkb) [inline views/views.py]: Bad practice to mark block end. Instead mark the start of the next block if that might be confusing.

File: views/views.py [hunk 17]
```diff
@@ -147,6 +262,50 @@ def create(self, request, *args, **kwargs):
             comment = serializer.save(commenter=person)
             query.comments.add(comment)
             query.save()
+
+            # Notifications start
+
```
Comment (dhruvkb) [inline views/views.py]: Drop the newline between a block start comment and the block itself.

File: views/views.py [hunk 18]
```diff
@@ -88,6 +97,112 @@ def partial_update(self, request, *args, **kwargs):
                     },
                     status=status.HTTP_403_FORBIDDEN,
                 )
+
+        # Notifications start
+
```
Comment (dhruvkb) [inline views/views.py]: Drop the newline between a block start comment and the block itself.

File: views/views.py (hunk 16 above)
Comment (shreyansh23) [reply views/views.py]: Okay, thanks I was not aware of that.

**Pull request merged**